lib/msal-core/src/UserAgentApplication.ts

```diff
@@ -388,201 +378,529 @@ export class UserAgentApplication {
+   * @param {string} extraQueryParameters - Key-value pairs to pass to the STS during the  authentication flow.
+   * @returns {Promise.<string>} - A Promise that is fulfilled when this function has completed, or rejected if an error was raised. Returns the token or error.
+   */
+  // TODO: params to accept AuthRequest object instead
+  acquireTokenPopup(scopes: Array<string>): Promise<string>;
+  acquireTokenPopup(scopes: Array<string>, authority: string): Promise<string>;
+  acquireTokenPopup(scopes: Array<string>, authority: string, user: User): Promise<string>;
+  acquireTokenPopup(scopes: Array<string>, authority: string, user: User, extraQueryParameters: string): Promise<string>;
+  acquireTokenPopup(scopes: Array<string>, authority?: string, user?: User, extraQueryParameters?: string): Promise<string> {
+    return new Promise<string>((resolve, reject) => {
+      // Validate scopes
+      // TODO: Replace with new validation pattern
+      const isValidScope = this.validateInputScope(scopes);
+      if (isValidScope && !Utils.isEmpty(isValidScope)) {
+        // TODO: Should reject with custom error
+        // TODO: Is this always accessToken? 
+        reject(ErrorCodes.inputScopesError + Constants.resourceDelimeter + isValidScope);
+      }
+
+      if (scopes) {
+        scopes = this.filterScopes(scopes);
+      }
+
+      const scope = scopes.join(" ").toLowerCase();
+
+      // Get the user object if session exists
+      const userObject = user ? user : this.getUser();
+
+      // If already in progress, reject the request
+      if (this._acquireTokenInProgress) {
+        // TODO: Should reject with custom error
+        reject(ErrorCodes.acquireTokenProgressError + Constants.resourceDelimeter + ErrorDescription.acquireTokenProgressError);
+        return;
+      }
+
+      
+      //if user is not currently logged in and no login_hint is passed
+      if (!userObject && !(extraQueryParameters && (extraQueryParameters.indexOf(Constants.login_hint) !== -1))) {
+        this._logger.info("User login is required");
+        reject(ErrorCodes.userLoginError + Constants.resourceDelimeter + ErrorDescription.userLoginError);
+        return;
+      }
+
+      // track the acquireToken progress
+      this._acquireTokenInProgress = true;
+
+      let authenticationRequest: AuthenticationRequestParameters;
+      let acquireTokenAuthority = authority ? AuthorityFactory.CreateInstance(authority, this.validateAuthority) : this.authorityInstance;
+      
+      // Open the popup window
+      var popUpWindow = this.openWindow("about:blank", "_blank", 1, this, resolve, reject);
+      if (!popUpWindow) {
+        // TODO: we should be rejecting with an error here
+        return;
+      }
+
+      acquireTokenAuthority.ResolveEndpointsAsync().then(() => {
+        // On fullfillment
+        if (Utils.compareObjects(userObject, this.getUser())) {
```

Comment:
compare object is hard to read... is it saying the two are equivalent?

Reply:
This should probably be renamed to compareUsers, but it is simply checking whether the userIdentifiers are equal to determine object equality, as any simple comparator would do. Please review the Utils file.